typing/typemod.ml

```diff
@@ -491,37 +490,45 @@ module Merge = struct
         let error = With_cannot_remove_packed_modtype(p,mty) in
         raise (Error(loc,initial_env,error))
 
-  let merge_constraint_aux initial_env loc sg lid constr : merge_result =
-    let destructive_substitution =
+  (* After the item has been patch, post processing does the actual destructive
+     substitution and checks wellformedness of the resulting signature *)
+  let post_process ~destructive loc lid env paths sg replace =
+    let sg =
+      if destructive then
+        (* Check that the substitution will not make the signature ill-formed *)
+        let _ = check_usage_after_substitution ~loc ~lid env paths sg in
+        (* Actually remove the identifiers *)
+        let sub = Subst.change_locs Subst.identity loc in
+        let sub = List.fold_left replace sub paths in
+        unsafe_signature_subst sub sg loc env
+      else sg
+    in
+    (* check that the resulting signature is still wellformed *)
+    let _ = check_well_formed_module env loc "this instantiated signature"
```

Comment:
```
let () = ...
```
to avoid suggesting that `check_well_formed_module` returns some ignored contents.